apis/v1alpha1/types.go

```diff
@@ -24,8 +24,15 @@ import (
 
 // A ProviderConfigSpec defines the desired state of a Provider.
 type ProviderConfigSpec struct {
-	// Credentials required to authenticate to this provider.
+	// Credentials used to connect to the Kubernetes API. Typically a
+	// kubeconfig file. Use InjectedIdentity for in-cluster config.
 	Credentials ProviderCredentials `json:"credentials"`
+
+	// Identity used to authenticate to the Kubernetes API. The identity
+	// credentials can be used to supplement kubeconfig 'credentials', for
+	// example by configuring a bearer token source such as OAuth.
+	// +optional
+	Identity *Identity `json:"identity,omitempty"`
```

Comment:
Currently these changes make a ProviderConfig look like this:


```yaml
apiVersion: helm.crossplane.io/v1beta1
kind: ProviderConfig
metadata:
  name: default
spec:
  credentials:
    source: Secret
    secretRef:
      name: cluster-credentials
      namespace: crossplane-system
      key: kubeconfig
  identity:
    type: GoogleApplicationCredentials
    source: Secret
    secretRef:
      name: gcp-credentials
      namespace: crossplane-system
      key: credentials.json
```

This is probably the safest and most backward compatible API, but I don't like it for a few reasons. Namely:

1. `credentials` and `identity` sound very similar. It's not obvious what is for what.
2. `credentials` is a misnomer for a kubeconfig, which configures more than creds, and may configure no creds.
3. I could imagine wanting to configure this provider (and others like provider-kubernetes) by just providing an API server URL etc without a full kubeconfig.

I think we could address my concerns above by making a larger change to the ProviderConfig API. This change would still be backward compatible, but in a kind of awkward way. It would instead look something like this:

```yaml
apiVersion: helm.crossplane.io/v1beta1
kind: ProviderConfig
metadata:
  name: default
spec:
  # 'kubernetes' is a new optional field.
  kubernetes:
    # Nesting kubeconfig here gives us the opportunity to specify Kubernetes configuration
    # differently in future - e.g. by loading an API server URL, CA cert, and token from different
    # connection detail keys.
    kubeconfig:
      # We support loading a kubeconfig from a Secret, env var or filesystem like we do for creds.
      source: Secret
      secretRef:
        namespace: crossplane-system
        name: cluster-creds
        key: credentials.json
  # 'credentials' becomes an optional field.
  # You must specify at least one of 'kubernetes' and 'credentials'.
  credentials:
    # The default type of credentials is Kubeconfig for backward compatibility, but we recommend
    # that folks use the 'kubernetes' field instead of Kubeconfig "credentials" type. The
    # 'credentials' field then becomes mostly a way to override or extend any credentials loaded
    # from a kubeconfig.
    type: GoogleApplicationCredentials
    source: Secret
    secretRef:
      namespace: crossplane-system
      name: cluster-creds
      key: credentials.json
```

That all said, I still don't love the above API because it seems like 'credentials' should be under 'kubernetes'.

Reply:
> Not sure whether there is a valid use case for this (don't remember such a request from community)

Right - I'm not aware of any requests from the community either so it's definitely hypothetical. My thinking is that I could conceive of a managed resource that represents a Kubernetes cluster that just returned keys for things like API server URL and CA cert data, rather than forming a whole kubeconfig. This is actually how some of our cluster managed resources used to work waaaay back around Crossplane v0.2 or v0.3 before we standardised them all on exposing a fully formed kubeconfig.